pkg/cnab/provider/parameters_test.go

```diff
@@ -179,16 +191,113 @@ func Test_loadParameters_fileParameter(t *testing.T) {
 		"foo": "/path/to/file",
 	}
 
-	params, err := r.loadParameters(b, overrides, "action")
+	args := ActionArguments{
+		Action: "action",
+		Params: overrides,
+	}
+	params, err := r.loadParameters(b, args)
 	require.NoError(t, err)
 
 	require.Equal(t, "SGVsbG8gV29ybGQh", params["foo"], "expected param 'foo' to be the base64-encoded file contents")
 }
 
+func Test_loadParameters_ParameterSourcePrecedence(t *testing.T) {
+	r := NewTestRuntime(t)
+	r.TestParameters.AddTestParameters("testdata/paramset.json")
+	r.TestParameters.TestSecrets.AddSecret("foo_secret", "foo_set")
+
+	r.TestConfig.TestContext.AddTestFile("testdata/bundle-with-param-sources.json", "bundle.json")
+	b, err := r.ProcessBundle("bundle.json")
+	require.NoError(t, err, "ProcessBundle failed")
+
+	overrides := map[string]string{
+		"foo": "foo_override",
+	}
+
+	t.Run("nothing present, use default", func(t *testing.T) {
+		args := ActionArguments{
+			Installation: "mybun",
+			Action:       claim.ActionUpgrade,
+		}
+		params, err := r.loadParameters(b, args)
+		require.NoError(t, err)
+		assert.Equal(t, "foo_default", params["foo"],
+			"expected param 'foo' to have default value")
+	})
+
+	t.Run("only override present", func(t *testing.T) {
+		args := ActionArguments{
+			Installation: "mybun",
+			Action:       claim.ActionUpgrade,
+			Params:       overrides,
+		}
+		params, err := r.loadParameters(b, args)
+		require.NoError(t, err)
+		assert.Equal(t, "foo_override", params["foo"],
+			"expected param 'foo' to have override value")
+	})
+
+	t.Run("only parameter source present", func(t *testing.T) {
+		c := r.TestClaims.CreateClaim("mybun", claim.ActionInstall, b, nil)
+		cr := r.TestClaims.CreateResult(c, claim.StatusSucceeded)
+		r.TestClaims.CreateOutput(c, cr, "foo", []byte("foo_source"))
+
+		args := ActionArguments{
+			Installation: "mybun",
+			Action:       claim.ActionUpgrade,
+		}
+		params, err := r.loadParameters(b, args)
+		require.NoError(t, err)
+		assert.Equal(t, "foo_source", params["foo"],
+			"expected param 'foo' to have parameter source value")
+	})
+
+	t.Run("override > parameter source", func(t *testing.T) {
+		c := r.TestClaims.CreateClaim("mybun", claim.ActionInstall, b, nil)
+		cr := r.TestClaims.CreateResult(c, claim.StatusSucceeded)
+		r.TestClaims.CreateOutput(c, cr, "foo", []byte("foo_source"))
+
+		args := ActionArguments{
+			Installation: "mybun",
+			Action:       claim.ActionUpgrade,
+			Params:       overrides,
+		}
+		params, err := r.loadParameters(b, args)
+		require.NoError(t, err)
+		assert.Equal(t, "foo_override", params["foo"],
+			"expected param 'foo' to have parameter override value")
+	})
+
+	t.Run("merge parameter values", func(t *testing.T) {
+		// foo is set by a the user
+		// baz is set by a parameter source
```

Comment:
nit: these appear to be swapped.  `baz` looks to be set by the default below, while `bar` is set by the parameter source.